File: src/chapters/linear-systems/content/code/forward-substitution.cpp
#include <vector>
#include <iostream>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Solve a lower-triangular system L y = b.
Vec forward_substitution(const Mat& L, const Vec& b) {
    int n = b.size();
    Vec y(n);
    for (int i = 0; i < n; ++i) {
        double s = b[i];
        for (int j = 0; j < i; ++j) s -= L[i][j] * y[j];
        y[i] = s / L[i][i];
    }
    return y;
}

int main() {
    Mat L = {{2, 0, 0}, {1, 3, 0}, {-1, 1, 2}};
    Vec b = {4, 5, -1};
    for (double v : forward_substitution(L, b)) cout << v << " ";
    cout << "\n";
}

File: src/chapters/linear-systems/content/code/forward-substitution.rs
// Solve a lower-triangular system L y = b.
fn forward_substitution(l: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = b.len();
    let mut y = vec![0.0; n];
    for i in 0..n {
        let mut s = b[i];
        for j in 0..i { s -= l[i][j] * y[j]; }
        y[i] = s / l[i][i];
    }
    y
}
fn main() {
    let l = vec![vec![2.0, 0.0, 0.0], vec![1.0, 3.0, 0.0], vec![-1.0, 1.0, 2.0]];
    let b = vec![4.0, 5.0, -1.0];
    println!("{:?}", forward_substitution(&l, &b));   // [2, 1, 0]
}
